content/draft/general/password_strength.md

```diff
@@ -0,0 +1,91 @@
+
+## Designs
+
+* [designs](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=11%3A4921)
+
+## Use Cases
+
+### Account creation
+
+1. User types the choosen password in the password field
+2. System displays feedback on the strength of the password
+2.1 System checks if the password fits criteria (see Functional Requirements#password criteria) and displays feedback (e.g "Password is too short")
+2.2 System categorizes password strength (see Functional Requirements#password categorization) and displays matching category (e.g password Is "Good")
+2.2 System checks if password is a common password (see Functional Requirements#password validation) and displays feedback (e.g "this pasword has been pwned and shouldn't be used")
+3. User types password again in the confirm password field
+4. System checks if password matches and displays "Passwords don't match" if they don't
+5. System checks if password matches and enables the "Create" button if they do
+6. User clicks "Create" button
+
+### Change password
+
+1. User types the current password in the current password field
+2. User types the choosen password in the new password field
+3. System displays feedback on the strength of the password
+4.1 System checks if the password fits criteria (see Functional Requirements#password criteria) and displays feedback (e.g "Password is too short")
+4.2 System categorizes password strength (see Functional Requirements#password categorization) and displays matching category (e.g password Is "Good")
+4.2 System checks if password is a common password (see Functional Requirements#password validation) and displays feedback (e.g "this pasword has been pwned and shouldn't be used")
+5. User types password again in the confirm password field
+6. System checks if password matches and displays "Passwords don't match" if they don't
+7. System checks if password matches and enables the "Create" button if they do
+8. User clicks "Create" button
+7. System checks if current password is valid and displays "Current password is invalid" if it isn't
+
+## Functional Requirements
+
+### Password criteria
+- A password MUST BE at least 6 characters long.
+- A password MUST BE at most 64 characters long.
+- A password MUST NOT contain spaces.
+- A password MAY CONTAIN any printable ascii character.
+
+### Password validation
+
+- A password MUST be checked against a list of common passwords
+    - If a password matches a common password, it MUST BE rejected
+    - If a password matches a common password, it MUST BE display a warning "this pasword has been pwned and shouldn't be used"
+    - If the application is online, the password MAY BE checked against an online or a local database.
```

Comment:
Is it really necessary to check the password against an online service? It doesnt feel safe to have the password leave your device.

Reply:
A local db would indeed be better, I don't think it would take much more spaces (only text) but probably something to confirm

Reply:
> Is it really necessary to check the password against an online service? It doesnt feel safe to have the password leave your device.

I guess the idea would be to **fetch** an online db and check against that locally. No need to send the PW over the wire

Reply:
the proposal is to use https://haveibeenpwned.com/API/v3#PwnedPasswords

offline is also ok, but can be quite limited, some of these databases are huge

Reply:
> 

with the https://haveibeenpwned.com/API/v3#PwnedPasswords service, the password is never sent over the wire

Reply:
Checking against previously breached passwords is indeed part of [the basic controls recommended by the NIST](https://github.com/usnistgov/800-63-3/blob/46a65dee0c99c9216ea994186748a19b23c64ae3/sp800-63b/sec5_authenticators.md?plain=1#L38)
> * Passwords obtained from previous breach corpuses.
> * Dictionary words.
> * Repetitive or sequential characters (e.g. 'aaaaaa', '1234abcd').
> * Context-specific words, such as the name of the service, the username, and derivatives thereof.

Although history has proven their recommendations should not necessarily be followed blindly, the first 3 items make a lot of sense today.

The question is should the Status app rely on a third party services for this ?
As our users machines will directly communicate with the service this should probably be an opt-in.

Also, a (weaker) alternative would be to check locally against a smaller database like https://github.com/danielmiessler/SecLists/blob/master/Passwords/Common-Credentials/10-million-password-list-top-1000000.txt